File: app/src/main/java/com/snap/model/shading/ShadingProgram.java
package com.snap.model.shading;

import android.opengl.GLES20;

import com.snap.model.exception.GlException;
import com.snap.model.exception.GlLibException;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Represents Pair of shaders - Vertex and Fragment.
 */
public class ShadingProgram {

    private static final int NOT_DEFINED_PROGRAM_HANDLE = 0;

    /**
     * Handle (reference) to shading program.
     */
    private int mProgramHandle;

    private final ShadingPair mShadingPair;

    public static int createShadingProgram(ShadingPair shadingPair) throws GlLibException {
        final int shadingProgramHandle = GLES20.glCreateProgram();
        if (shadingProgramHandle == NOT_DEFINED_PROGRAM_HANDLE) {
            throw new GlException("Could not create shading program.");
        }
        shadingPair.setupAndAttachToProgram(shadingProgramHandle);
        GLES20.glLinkProgram(shadingProgramHandle);
        final int[] linkStatus = new int[1];
        GLES20.glGetProgramiv(shadingProgramHandle, GLES20.GL_LINK_STATUS, linkStatus, 0);
        if (linkStatus[0] != GLES20.GL_TRUE) {
            String errorMessage = GLES20.glGetProgramInfoLog(shadingProgramHandle);
            GLES20.glDeleteProgram(shadingProgramHandle);
            throw new GlException("Could not link shading program. " + errorMessage);
        }
        return shadingProgramHandle;
    }

    public ShadingProgram(ShadingPair shadingPair) {
        mShadingPair = shadingPair;
    }

    public boolean isSetup() {
        return mProgramHandle != NOT_DEFINED_PROGRAM_HANDLE;
    }

    public void setup() throws GlLibException {
        if (isSetup()) {
            throw new GlLibException("Shading program already setup.");
        }
        mProgramHandle = createShadingProgram(mShadingPair);
    }

    public void release() {
        if (isSetup()) {
            GLES20.glDeleteProgram(mProgramHandle);
            mProgramHandle = NOT_DEFINED_PROGRAM_HANDLE;
        }
    }

    public void useProgram() {
        if (!isSetup()) {
            throw new IllegalStateException("You have to setup program before using it.");
        }
        GLES20.glUseProgram(mProgramHandle);
    }

    public void executeAction(final ShadingProgramAction shadingProgramActionToExecute) {
        useProgram();
        shadingProgramActionToExecute.execute(this);
    }

    public AttributeBinding createAttributeBinding(final String attributeName) {
        return new AttributeBinding(mProgramHandle, attributeName);
    }

    public UniformBinding createUniformBinding(final String uniformName) {
        return new UniformBinding(mProgramHandle, uniformName);
    }

    public static class ShadingPair {

        private final Shader vertexShader;
        private final Shader fragmentShader;

        public ShadingPair(Shader vertexShader, Shader fragmentShader) {
            if (vertexShader.getType() != Shader.ShaderType.VERTEX) {
                throw new IllegalArgumentException("First param has to be a Vertex shader.");
            }
            if (fragmentShader.getType() != Shader.ShaderType.FRAGMENT) {
                throw new IllegalArgumentException("Second param has to be a Fragment shader.");
            }
            this.vertexShader = vertexShader;
            this.fragmentShader = fragmentShader;
        }

        public Shader getVertexShader() {
            return vertexShader;
        }

        public Shader getFragmentShader() {
            return fragmentShader;
        }

        void setupAndAttachToProgram(final int shadingProgramHandle) throws GlLibException {
            vertexShader.setup();
            GLES20.glAttachShader(shadingProgramHandle, vertexShader.getShaderHandle());
            fragmentShader.setup();
            GLES20.glAttachShader(shadingProgramHandle, fragmentShader.getShaderHandle());
        }
    }

    public interface ShadingProgramAction {

        void execute(ShadingProgram onProgram);
    }
}

File: app/src/main/java/com/snap/model/shading/ShadingProgram.kt
package com.snap.model.shading

import android.opengl.GLES20

import com.snap.model.exception.GlException
import com.snap.model.exception.GlLibException

private const val notDefinedProgramHandle = 0

/**
 * Represents Pair of shaders - Vertex and Fragment.
 */
internal class ShadingProgram(private val vertexShader: Shader,
                              private val fragmentShader: Shader) {

    init {
        if (vertexShader.type !== ShaderType.VERTEX) {
            throw IllegalArgumentException("First param has to be a Vertex shader.")
        }
        if (fragmentShader.type !== ShaderType.FRAGMENT) {
            throw IllegalArgumentException("Second param has to be a Fragment shader.")
        }
    }

    /**
     * Handle (reference) to shading program.
     */
    private var programHandle: Int = 0

    val isSetup: Boolean
        get() = programHandle != notDefinedProgramHandle

    fun setup() {
        if (isSetup) {
            throw GlLibException("Shading program already createAndCompile.")
        }
        programHandle = createShadingProgram(vertexShader, fragmentShader)
    }

    private fun createShadingProgram(vertexShader: Shader, fragmentShader: Shader): Int {
        val shadingProgramHandle = GLES20.glCreateProgram()
        if (shadingProgramHandle == notDefinedProgramHandle) {
            throw GlException("Could not create shading program.")
        }

        vertexShader.createAndCompile()
        GLES20.glAttachShader(shadingProgramHandle, vertexShader.shaderHandle)

        fragmentShader.createAndCompile()
        GLES20.glAttachShader(shadingProgramHandle, fragmentShader.shaderHandle)

        GLES20.glLinkProgram(shadingProgramHandle)
        val linkStatus = IntArray(1)
        GLES20.glGetProgramiv(shadingProgramHandle, GLES20.GL_LINK_STATUS, linkStatus, 0)
        if (linkStatus[0] != GLES20.GL_TRUE) {
            val errorMessage = GLES20.glGetProgramInfoLog(shadingProgramHandle)
            GLES20.glDeleteProgram(shadingProgramHandle)
            throw GlException("Could not link shading program. $errorMessage")
        }

        return shadingProgramHandle
    }

    fun release() {
        if (isSetup) {
            GLES20.glDeleteProgram(programHandle)
            programHandle = notDefinedProgramHandle
        }
    }

    fun useProgram() {
        if (!isSetup) {
            throw IllegalStateException("You have to createAndCompile program before using it.")
        }
        GLES20.glUseProgram(programHandle)
    }

    fun doUsingProgram(block: ShadingProgram.() -> Unit) = apply {
        useProgram()
        block()
    }

    fun createAttributeBinding(attributeName: String, block: AttributeBinding.() -> Unit = {}) =
            AttributeBinding(programHandle, attributeName).apply(block)

    fun createUniformBinding(uniformName: String, block: UniformBinding.() -> Unit = {}) =
            UniformBinding(programHandle, uniformName).apply(block)
}
